Preserve user comments in bib file

lenhard opened the pull request:
So I am finally having my take at #1026. The current solution was surprisingly easy so far (despite what I had written before). It looks explicitly for our encoding prefix in the file ("Encoding: ") and kills lines that contain it (but only lines preceding an entry, it will not delete something that is inside an entry). Other user comments are left untouched and are *always* written out again, even if the entry is reformatted.

User comments that are above meta data, bibtex strings, or the preamble will still be removed, though. Changing that would require large scale changes to these items in our model, since we would also need to store the parsed serialization for them, which we do not do so far.

This should receive significant automated, but also manual, testing, since it modifies a quite critical part and has the potential to destroy bib files. I will add a few more tests for this PR.

- [X] Change in CHANGELOG.md described
- [X] Tests created for changes

lenhard commented:
Ok, seems like rebasing my stuff here on master was not the smartest thing to do...

lenhard commented:
Ok, I'd like to put this up for discussion.

User comments are now kept under the following circumstances:
 - Above any BibEntry
 - Above any BibtexString
 - Comments at the end of the file

This is independent of whether an entry is changed/reformatted, but not if it is deleted. If it is reformated, user comment text is located exactly one blank line above the reformated entry.

User comments are still not kept when:
 - Above meta data comments
 - Above the Preamble
 - Contains JabRef's ENCODING_PREFIX
 - Above an entry that has been deleted

Is this good enough? It should be sufficient in most scenarios and work if someone opens his non-JabRef bib file with JabRef. If someone starts adding arbitrary text between JabRef's Metadata, though... The latter is hard to achieve since we do not store the serialization of MetaData. If we want to, I'd have to make some additions to our MetaData objects. I could also store the comments between meta data and add them as a bunch at the end of the file, but then the relative position gets lost. Before I do more work, I'd like to clarify the following points:
 - Does the storage of arbitrary text above MetaData really matter?
 - Is it ok to delete comments if the entry below them is deleted?

tobiasdiez commented on CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 ## [Unreleased]
 
 ### Changed
+- JabRef does no longer delete user comments outside of BibTeX entries [#1026]
```
Change format of changelog entry to match the other ones.

tobiasdiez commented on src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java:
```diff
@@ -354,6 +356,27 @@ private void writeString(Writer fw, BibtexString bs, Map<String, BibtexString> r
+
+            try {
+                // get the text before the string
+                String prolog = string.getParsedSerialization().substring(0, string.getParsedSerialization().indexOf('@'));
```
Idea: Move most of the code which determines the comment to `BibEntry.getUserComments` since in my opinion it is the responsibility of the BibEntry to store all the data in an accessible format.

lenhard replied:
Good idea, will do.

However, I am a little bit opposed to simplifying the roundtrip test. This is the discussion on whether we prefer unit over integration tests. We have been there before :)

tobiasdiez commented on src/main/java/net/sf/jabref/logic/bibtex/BibEntryWriter.java:
```diff
@@ -39,22 +39,46 @@ public void write(BibEntry entry, Writer out, BibDatabaseMode bibDatabaseMode) t
+
+            try {
+                // get the text before the entry
+                String prolog = entry.getParsedSerialization().substring(0, entry.getParsedSerialization().indexOf('@'));
```
Same as above: move to `BibEntry`

tobiasdiez commented on src/test/java/net/sf/jabref/importer/fileformat/BibtexParserTest.java:
```diff
@@ -1253,10 +1254,10 @@ public void parseSavesNewlinesBeforeEntryInParsedSerialization() throws IOExcept
         String testEntry = "@article{test,author={Ed von Test}}";
         ParserResult result = BibtexParser.parse(
-                new StringReader("%Encoding: no" + Globals.NEWLINE + Globals.NEWLINE + Globals.NEWLINE + testEntry));
+                new StringReader(Globals.ENCODING_PREFIX + Globals.NEWLINE + Globals.NEWLINE + Globals.NEWLINE + testEntry));
```
the encoding "no" is missing in the new version

lenhard replied:
That does not matter for the test (so in a sense, I minimized the test ;-) ) Only the `ENCODING_PREFIX` is relevant. A line with it will be purged regardless of what encoding is set.

tobiasdiez commented on src/test/java/net/sf/jabref/importer/fileformat/BibtexParserTest.java:
```diff
@@ -1486,4 +1487,43 @@ public void parseReturnsEntriesInSameOrder() throws IOException {
+        ParserResult result = BibtexParser.parse(new StringReader(bibtexEntry));
+
+        Collection<BibEntry> entries = result.getDatabase().getEntries();
+        assertEquals(1, entries.size());
```
Create expected bibentry and compare against it.

lenhard replied:
I do not really understand your comment. The last line in the test method compares the expected serialization with the actual one: `        assertEquals(bibtexEntry, entry.getParsedSerialization());`

tobiasdiez replied:
I meant:
````
BibEntry expected = new BibEntry().withField("author", "Foo Bar).withField(.....)
assertEquals(Collections.singletonList(expected), entries)
````

lenhard replied:
The `withField` syntax is something new for me.

Anyway, this will not work here, unfortunately. The crucial part is the comment before the entry and this cannot be set as a field.

tobiasdiez replied:
So what you actually want to test is
assertEquals("% Some random comment that should stay here", entry.getUserComment());
right?

tobiasdiez commented on src/test/java/net/sf/jabref/logic/bibtex/BibEntryWriterTest.java:
```diff
@@ -403,4 +403,83 @@ public void trimFieldContents() throws IOException {
+        assertEquals(1, entries.size());
+
+        BibEntry entry = entries.iterator().next();
+        assertEquals("test", entry.getCiteKey());
```
Remove these asserts

lenhard replied:
Why?

tobiasdiez replied:
Since a test should only fail because of one reason (i.e. in the best case only one assert). This test should check if parsing -> writing yields the same string which was originally passed to it. Thus only the last assert is what you want. https://github.com/JabRef/jabref/pull/1471/files#diff-b4b7ca399d396eb3e62fc3416156c92aR437

lenhard replied:
Ok, I went through the class and removed all asserts I found redundant.

tobiasdiez commented on src/test/java/net/sf/jabref/importer/fileformat/BibtexParserTest.java:
```diff
@@ -1486,4 +1487,43 @@ public void parseReturnsEntriesInSameOrder() throws IOException {
+    @Test
+    public void parsePrecedingComment() throws IOException {
+        // @formatter:off
+        String bibtexEntry = "% Some random comment that should stay here" + Globals.NEWLINE +
```
What happens to some random test in front of the entry? Bib(la)tex does not know the percentage symbol as a comment symbol...

lenhard replied:
This does not make a difference. Any non-whitespace text (except for the `ENCODING_PREFIX`) is treated as a comment. In a sense, the percentage sign here is a complete coincidence.

tobiasdiez commented on src/test/java/net/sf/jabref/importer/fileformat/BibtexParserTest.java:
```diff
@@ -1486,4 +1487,43 @@ public void parseReturnsEntriesInSameOrder() throws IOException {
+    public void parsePrecedingComment() throws IOException {
+        // @formatter:off
+        String bibtexEntry = "% Some random comment that should stay here" + Globals.NEWLINE +
+                "@Article{test," + Globals.NEWLINE +
```
Add test for entries starting on the same line as the comment, i.e. `some text containing % and now the @article {...}`

tobiasdiez commented on src/test/java/net/sf/jabref/logic/bibtex/BibEntryWriterTest.java:
```diff
@@ -403,4 +403,83 @@ public void trimFieldContents() throws IOException {
+        Collection<BibEntry> entries = result.getDatabase().getEntries();
+        assertEquals(1, entries.size());
+
+        BibEntry entry = entries.iterator().next();
```
Also remove asserts here

lenhard replied:
As above: why?

tobiasdiez commented on src/test/resources/testbib/bibWithEpilog.bib:
```diff
@@ -0,0 +1,43 @@
+  location = {Shanghai, China}
+}
+
+@Comment{jabref-meta: databaseType:bibtex;}
```
I don't like these files for round-trips. In my opinion it would suffice to add comments to `complex.bib` which already has round-trip tests (as integration test) and concentrate on unit test for the rest.

lenhard replied:
I tend to disagree. If I point all kinds of test to `complex.bib`, these tests are indirectly coupled to each other. I would rather like to keep them as separate as possible and that means separate files.

tobiasdiez replied:
I think you missunderstood me. For complex.bib there are already round-trip tests in place. Thus you just need to modify `complex.bib` instead of writing a new round-trip test with a new bib file.

lenhard replied:
I see. And `complex.bib` already has an epilog. That means the above test can simply be removed.

tobiasdiez commented:
Looks good to me. Just had some more or less minor remarks (mostly about the testing code). 

I think we don't need support for comments between JabRef metadata.

lenhard commented:
@tobiasdiez If not mentioned otherwise, I addressed your comments above.

We tend to differ when it comes to testing in general, though. I really prefer integration tests when doing anything with reading/writing. In JabRef, these things are inherently coupled. To implement this PR (write out user comments correctly), I also had to do significant changes to reading. So, as long as no overwhelming majority of @JabRef/developers pressures me to write unit tests, I will rather write integration tests.

simonharrer commented:
LGTM. I think integration tests are more valuable. So this would suffice for me.

lenhard commented:
Great! Let us talk briefly about this during the next devcall and decide if we take it into 3.5 or 3.6.

tobiasdiez commented:
For 3.6

koppor commented on jabref.install4j:
```diff
@@ -369,7 +369,7 @@ return true;</string>
                 <condition>context.getBooleanVariable("addToDockAction")</condition>
               </action>
-              <group name="Registry" id="239" customizedId="" beanClass="com.install4j.runtime.beans.groups.ActionGroup" enabled="true" commentSet="false" comment="" actionElevationType="inherit">
+              <group name="Registry" id="239" customizedId="" beanClass="com.install4j.runtime.beans.groups.ActionGroup" enabled="true" commentSet="false" comment="" actionElevationType="elevated">
```
Why is an install4j change part of this PR?

stefan-kolb replied:
I revert it later, sorry...

tobiasdiez commented on src/main/java/net/sf/jabref/model/entry/BibtexString.java:
```diff
@@ -152,4 +152,28 @@ public String getParsedSerialization() {
+    /*
+    * Returns user comments (arbitrary text before the string) if there are any. If not returns the empty string
+     */
+    public String getUserComments() {
```
This is now an exact copy of BibEntry.getUserComments right? Maybe add a superclass BibItem containing this method and then BibEntry and BibString derive from this superclass (maybe there is even more common code which could be extracted to the super class)

simonharrer replied:
Sorry to always slip into the role of your antagonist, but I am against creating a class hierarchy. In most cases, it makes the code harder to understand, and the code duplication savings are just not worth the effort. In addition, one has to adhere to the liskov substitution principle to create a good class hierarchy, but this is hard to do right.

tobiasdiez replied:
No problem, having an experienced antagonist is probably the best way to learn something :smile: 

So now ignoring everything about the actual implementation and only speaking about "business objects": There are different items which can be contained in a bib file. For example `@Comments`, `@Strings` and normal bib entries. They all have a similar structure (e.g begin with `@`, followed by some identifier and then braces which surround the actual content) and all of them could have some text comments in front of them. Thus there is also a common behavior when it comes to parsing and writing.  
So how would my deer antagonist reflect this common structure and behavior in java code?

My idea would be: BibString, BibEntry, BibComment all derive from BibItem. BibItem has methods to parse and write at least the form `@something { abstract method to parse / write value }`. Then the writer just gets a list of BibItems and invokes the write method on them. Similarly the parser returns a list of BibItems while the parsing of every item was done in the subclass. But again...this idea somehow corresponds from a `service` implementation to a full-blown domain model....so yeah, maybe we should discuss this in the next dev call ;-)

koppor commented on CHANGELOG.md:
```diff
@@ -13,12 +13,6 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 
 ### Changed
 - [#1026](https://github.com/JabRef/jabref/issues/1026) JabRef does no longer delete user comments outside of BibTeX entries and strings
-- Implemented [#661](https://github.com/JabRef/jabref/issues/661): Introducing a "check for updates" mechnism (manually/automatic at startup)
```
Why are these CHANGELOG.md entries deleted?

tobiasdiez commented on src/test/java/net/sf/jabref/exporter/BibDatabaseWriterTest.java:
```diff
@@ -292,6 +292,82 @@ public void roundtrip() throws IOException {
+    }
+
+    @Test
+    public void roundtripWithUserCommentBeforeString() throws IOException {
```
This looks like an exact copy of https://github.com/JabRef/jabref/blob/304d2802162a7e75edda5fbc893666cc73efe669/src/test/java/net/sf/jabref/exporter/BibDatabaseWriterTest.java#L279

stefan-kolb commented:
@lenhard Please squash the commits via Github when merging this :smile:

tobiasdiez commented on src/test/java/net/sf/jabref/logic/bibtex/BibEntryWriterTest.java:
```diff
@@ -117,14 +117,8 @@ public void roundTripWithPrependingNewlines() throws IOException {
-        assertEquals(1, entries.size());
 
         BibEntry entry = entries.iterator().next();
-        assertEquals("test", entry.getCiteKey());
```
Why did you removed this (suboptimal) assertions?

lenhard replied:
Are you being serious? In this very same PR you asked me to reduce the number asserts as far as possible, aiming for one assertion per test. This is the reason.

tobiasdiez commented on src/test/java/net/sf/jabref/logic/bibtex/BibEntryWriterTest.java:
```diff
@@ -149,18 +143,15 @@ public void roundTripWithModification() throws IOException {
         Set<String> fields = entry.getFieldNames();
         assertTrue(fields.contains("author"));
-        assertEquals("BlaBla", entry.getField("author"));
+        assertEquals("BlaBla", entry.getFieldOptional("author").get());
```
assertEquals(Optional.of(BlaBla), entry.getFieldOptional("author"))

lenhard replied:
Implemented
